tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,357 @@
+/**
+ * @file   mem_filesystem.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2020 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements the in-memory filesystem class
+ */
+
+#include <sstream>
+#include <unordered_set>
+
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/filesystem/mem_filesystem.h"
+#include "tiledb/sm/global_state/global_state.h"
+#include "tiledb/sm/misc/utils.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+class MemFilesystem::FSNode {
+ public:
+  FSNode(const std::string& name)
+      : name_(name){};
+
+  virtual ~FSNode() = default;
+
+  std::string get_name() const {
+    return name_;
+  }
+
+  virtual bool is_dir() const = 0;
+  virtual bool is_file() const = 0;
+  virtual std::vector<std::string> ls(const std::string& dir) const = 0;
```

Comment:
I don't think this needs an argument. Ideally, it should only return its children if it is a directory. Maybe the argument should be `bool recursive` that will return files in the child subdirectories.

Reply:
Sorry, bad naming. "dir" actually refers to the full path to this directory/file. The reason I added this argument is that from what I saw in VFS, ls is expected to return full paths to files/directories and not just the filenames. Eg. MemFilesystem.ls("/tmp/foo/") returns ["/tmp/foo/a.txt", "tmp/foo/bar"] instead of ["a.txt", "bar"].
If that way looks too hack-y to you, I can try to add a parent path to all FSNodes as an attribute, but this will require some extra trouble, especially in case of move/copy operations.

Reply:
Sounds good to me. I would still add a `recursive` argument and use it to implement the `MemFileSystem::ls()` this way:

```
FSNode* lookup_node(
    FSNode* current_node,
    vector<string>::iterator path_iter,
    vector<string>::iterator path_iter_end) {

  // Return the current node if we have successfully parsed all path tokens.
  if (path_iter == path_iter_end)
   return current_node;

  // Return a nullptr if the parent does not contain the next path token.
  const string& path_token = *path_iter;
  unordered_map<string, FSNode*>::iterator children_iter = current_node->children_.find(path_token);
  if (children_iter == parent->children_.end())
    return nullptr;

  FSNode *child_node = children_iter->second;
  return lookup_node(child_node, ++path_iter, path_iter_end);
}

FSNode* lookup_node(string path) {
  vector<string> path_tokens = tokenize(path);
  return lookup_node(path_tokens.begin(), path_tokens.end());
}

Status MemFileSystem::ls(string path, std::vector* child_paths) {
  FSNode *node = lookup_node(path);
  if (node == nullptr)
    return ...;

  *child_paths = node->ls(path, true /* recursive */);

  return Status::Ok();
}
```

Reply:
I keep this comment open for now and I will address it a bit later, thanks

Reply:
I was thinking that this is not needed for now, as there is no test for it I am suggesting to add this extra functionality when needed in the future together with a corresponding test. Is that ok ?